shared/src/main/scala/pl/metastack/metatime/Date.scala

```diff
@@ -75,17 +68,120 @@ trait Date extends Year with Month with Day {
     Year(y).unix().value +
     Month(m).unix().value +
     Day(d).unix().value)
-  }
+
+}
 
 object Date {
-  def now(): Date = ???
-  def apply(year: Int): Date = ???
+  def apply(year: Int): Date = new Date {
+    override val y: Int = year
+    override val m: Int = 1
+    override val d: Double = 1.0
+  }
   def apply(year: Int, month: Int): Date = ???
   def parse(format: String, timezone: Timezone): Date = ???
 
-  def apply(year: Int = 1970,
+  def findYearsOrRemainingDaysOfYear(totalDays: Int, year: Boolean) : Int = {
+    if(year) { yearsPassed(totalDays)}
+    else { daysPassed(totalDays) }
+  }
+
+  val YearsSequence = Seq(365, 365, 366, 365)
+
+  def yearsPassed(totalDays: Int) : Int = {
+    val yearLength = Stream.continually(YearsSequence).flatten.take((totalDays / 365) + 1).toList
+    val accumulatedDays = yearLength.zipWithIndex.foldLeft(Seq.empty[Int]) {
+      case (Nil, (cur, _)) => Seq(cur)
+      case (acc, (cur, i)) => acc ++ Seq(acc.last + cur)
+    }
+    accumulatedDays.zipWithIndex.indexWhere( {
+      case (accDays, year) => totalDays <= accDays
+    }) + 1970
+  }
+
+  def daysPassed(totalDays: Int) : Int = {
+    val yearLength = Stream.continually(YearsSequence).flatten.take((totalDays / 365) + 1).toList
+    val accumulatedDays = yearLength.zipWithIndex.foldLeft(Seq.empty[Int]) {
+      case (Nil, (cur, _)) => Seq(cur)
+      case (acc, (cur, i)) => acc ++ Seq(acc.last + cur)
+    }
+    accumulatedDays.zipWithIndex.indexWhere( {
+      case (accDays, year) => totalDays <= accDays
+    }) match {
+      case 0 => totalDays
+      case n => totalDays - accumulatedDays(n-1)
+    }
+  }
+
+  val DaysInMonths = Seq(31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
+
+  def isLeapYear(year: Int): Boolean = (year % 4) == 0
+
+
+  def findMonthOrDay(days: Int, year: Int, month: Boolean): Int = {
+    if(month) { findMonth(days, year)}
+    else { findDay(days, year) }
+  }
+
+  def findMonth(days: Int, year: Int): Int = {
+    val accumulatedDays = DaysInMonths.zipWithIndex.foldLeft(Seq.empty[Int]) {
```

Comment:
You can create a method `accumulatedDays` and use it in `findMonth` and `findDay`. Small, compositional functions increase testability.